File: labs/00-local-kafka/create-topics.js
import { kafka } from "../../lib/kafka.js";
import { topics } from "../../lib/topics.js";

const admin = kafka.admin();

try {
  await admin.connect();

  const existingTopics = new Set(await admin.listTopics());
  const missingTopics = topics.filter(({ topic }) => !existingTopics.has(topic));

  if (missingTopics.length === 0) {
    console.log("All lab topics already exist.");
  } else {
    await admin.createTopics({
      waitForLeaders: true,
      topics: missingTopics
    });

    console.log("Created topics:");
    for (const { topic, numPartitions } of missingTopics) {
      console.log(`- ${topic} (${numPartitions} partition(s))`);
    }
  }
} finally {
  await admin.disconnect();
}

File: labs/00-local-kafka/create_topics.py
from confluent_kafka import KafkaException

from lab_kafka.client import admin_client
from lab_kafka.topics import TOPICS


def main() -> None:
    admin = admin_client()
    existing_topics = set(admin.list_topics(timeout=10).topics)
    missing_topics = [topic for topic in TOPICS if topic.topic not in existing_topics]

    if not missing_topics:
        print("All lab topics already exist.")
        return

    futures = admin.create_topics(missing_topics)

    print("Creating topics:")
    for topic_name, future in futures.items():
        try:
            future.result()
            print(f"- {topic_name}")
        except KafkaException as error:
            print(f"- {topic_name}: {error}")


if __name__ == "__main__":
    main()

